include/circt/Dialect/Interop/Interop.td

```diff
@@ -42,4 +56,159 @@ def InteropDialect : Dialect {
+  }];
+}
+
+def ProceduralInitOp : InteropOp<"procedural.init", [SingleBlock]> {
+  let summary = "computes the initial values for the allocated state";
+  let description = [{
+    The `interop.procedural.init` operation takes the variadic list of states
+    from the `interop.procedural.alloc` operation as operands and has a body
+    with a `interop.return` operation that has a variadic list of operands that
+    matches the types of the states and represent the initial values to be
+    assigned to the state values.
+    The assignment will be inserted by the container-side lowering of the
+    interop operations.
+    The operation also has an interop mechanism attribute to allow bridging
+    patterns to map the types to valid types in another interop mechanism
+    and to wrap the operations in the body in a way to make them executable
+    in the other interop mechanism, e.g., wrap them in a `extern "C"` function
+    to make it callable from C or LLVM IR.
+  }];
+
+  let arguments = (ins InteropMechanismAttr:$interopMechanism,
+                       Variadic<AnyType>:$states);
+  let regions = (region SizedRegion<1>:$initRegion);
+
+  let assemblyFormat = [{
+    $interopMechanism ( $states^ `:` qualified(type($states)) )?
+    attr-dict-with-keyword $initRegion
+  }];
+
+  let builders = [
+    OpBuilder<(ins "ValueRange":$states, "InteropMechanism":$interopMechanism)>
+  ];
+
+  let hasVerifier = 1;
+  let skipDefaultBuilders = 1;
+}
+
+def ProceduralUpdateOp : InteropOp<"procedural.update", [
+  SingleBlock, AttrSizedOperandSegments
+]> {
+  let summary = "takes some persistent state and inputs to compute some results";
+  let description = [{
+    The `interop.procedural.update` operation has an interop mechanism attribute
+    to allow bridging patterns to map the types to valid types in another
+    interop mechanism and to wrap the operations in the body in a way to make
+    them executable using the other interop mechanism.
+
+    It takes the state values returned by the `interop.procedural.alloc`as
+    operands and passes them on to the body via block arguments using
+    pass-by-value semantics. In addition to the state values, it also takes a
+    variadic list of inputs and also passes them on to the body.
+    The `interop.return` inside the body then returns the result values after
+    doing some computation inside the body.
+
+    If the state needs to be mutated, it has to be a pointer type.
+  }];
+
+  let arguments = (ins InteropMechanismAttr:$interopMechanism,
+                       Variadic<AnyType>:$states,
+                       Variadic<AnyType>:$inputs);
+  let results = (outs Variadic<AnyType>:$outputs);
+  let regions = (region SizedRegion<1>:$updateRegion);
+
+  let assemblyFormat = [{
+    $interopMechanism ( ` ` `[` $states^ `]` )? ( `(` $inputs^ `)` )? `:`
+    (`[` qualified(type($states))^ `]`)? functional-type($inputs, $outputs)
+    attr-dict-with-keyword $updateRegion
+  }];
```

Comment:
It is a good day when everything can be done with the existing `assemblyFormat` and without a custom printer/parser.